### PR TITLE
feat(junit reporter): link testcases to Xray test issues and provide additional metadata for Xray Test Management

### PR DESCRIPTION
**Pull request by bitcoder**
Provides support for specific enhancements on the junit report, allowing deeper integration with Xray Test Management to show additional information on Jira side, or do some related operations upon the time of importing tesults.
The test metainformation may be specified using annotations on the `TestInfo` object.

This PR gives users the ability to:
- link a Playwright test to an existing Test in Xray, by referring its issue key using an annotation with type `test_key`
- link a Playwright test to an existing Test in Xray, by referring its issue id using an annotation with type `test_id`
- link a Playwright test to an existing story/requirement kinda of issue in Jira, by referring its key using an annotation with type `requirements`; multiple requirements may be specified, using comma as a delimiter on the annotation's description
- specify the Test issue' summary, to be used during provisioning of the Test in Xray if it doesn't exist yet, using an annotation with type 'test_summary'
- specify the Test issue's description, to be used during provisioning of the Test in Xray if it doesn't exist yet, using an annotation with type 'test_description'
- embed attachments, including its contents, on the report, by flagging it using a `embed_attachments_in_report` annotation with 'true' as its description
 
Resolves #9348

### CODE REVIEW

**Comment by btd** [comment]
I would say we need generic way to set properties in junit reporter, like described there: https://github.com/microsoft/playwright/issues/10942
As you can see in mentioned issue, a lot of different properties could be added. I do not think it is right idea to add support for specific tool, but just a generic way to set properties

**Comment by bitcoder** [comment]
@btd thanks for your feedback; I understand your point but this PR and feature request  is targeted for adding support for Xray Test Management and make that process as clean as possible for these users (by curiosity and to have just a more clear idea, Xray is used by several millions of users in more than 17K orgs). 
Making it more generic is possible *but* would making the integration with Xray more cumbersome; for example, with Xray the values of some properties are embed as "value" attribute, while others (because of their nature) are embed as inner cdata content. In the case of attachments, these have a specific mapping on the JUnit XML, with multiple inner `<item>` elements, with specific attributes and inner content. The current implementation hides this logic and makes it straighforward to use the attachments on the `TestInfo` object.
From what I saw, the datadog feature request #10942, also mentions `properties` at `testsuite` level, which the current PR doesn't address nor is supported by the built-in junit reporter on playwright-test.

As I've invested quite some time on the PR, I'll wait for feedback from the playwright team before doing additional changes.

**Comment by pdinhgia** [comment]
Hello, any updates about this issue ? I'm very interested about this feature

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [inline packages/playwright-test/src/reporters/junit.ts:143]: I am hesitant supporting only some annotations. Why don't we put all annotations as properties? Should be safe enough, as long as Xray ignores unknown properties, which it probably does?

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [inline packages/playwright-test/src/reporters/junit.ts:171]: Why do we have to put this one as text, and not as `value=` as others? I'd assume that these are interchangeable, it's just a matter of escaping.

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [inline packages/playwright-test/src/reporters/junit.ts:179]: I think it's better to have an option on the whole junit reporter, to say "embed attachments". Even in this case, we probably should not embed files, wdyt?

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [inline packages/playwright-test/src/reporters/junit.ts:185]: Let's add an option to the junit reporter: `{ embedAttachmentsAsProperty: 'testrun_evidence' }`? This seems a very specific name to put here.

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [inline packages/playwright-test/src/reporters/junit.ts:213]: When embedding attachments, we should skip the "reference attachments" logic on the line 245.

**Review by dgozman — COMMENTED** [review]
Sorry for being late! This PR looks very good overall, but I think we should make it a bit more generic - see comment. Thank you for contributing ❤️

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:296]: Well, I remember that whenever I was testing I wasn't getting some errors because it would produce invalid XML on the parser I was using. Nevertheless, if that block is character data it should be in a [CDATA block](https://www.w3.org/TR/REC-xml/#sec-cdata-sect). It doesn't make sense in my perspective, know that is character data and then not properly identify as such, right? :)
https://www.w3.org/TR/REC-xml/#sec-cdata-sect

Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:296]: It's a bit unrelated indeed but I think it should be the standard behaviour (I don't see a need to create an option for it unless there's a known use case where having CDATA is going to break anything for any parser out there, which I doubt but sometimes there are basic parsers...)

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:143]: Well, this annotations are the ones Xray understands as text based values. We could make it more generic (please see next comment)

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:185]: That `testrun_evidence` name is the one that Xray parses but I'm ok having it configurable, just let me know if you want to proceed that way.
Just for information, JUnit doesn't specify how to embed attachments in the report. This is something specified by the Xray team (even though I've validated it with JUnit5 team leader at high-level just to know if it would hurt the "schema" or go against some principles), that eventually other tools can take advantage of :)

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:213]: I agree

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:179]: Embeding attachments is in my perspective something that users will need to opt-in explicitly and should be disabled by default as it will make reports larger.
I'm not sure if I fully understood you, but please feel free to clairify/correct.
If you're suggesting to add a global configuration for the whole reporter, I agree, having it as false by default. I'm not sure if you were suggesting to remove this logic at the test level using the 'embed_attachments_in_report' annotation type that I've implemented.. or if you were suggesting having that global configuration plus this logic... or something else :)
Well, in my perspective maybe a global option for the junit reporter would be enough... not sure if users having that option as true, would want to disable this behavior in some test(s). If this makes sense, then the existing logic combined with the global option could make sense. What are your thoughts?

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:171]: For the time being, Xray processes some properties from a `value` attribute and other ones from inner text, due to their nature. It's not interchangeable.
So we have a set of well-known properties are are either mapped as "myproperty1" or as shown for "myproperty2".

`
<property name="myproperty1" value="dummyvalue"/>
<property name="myproperty2">
 a very long description .....
</property>
`

To make this mechanism generic and process either from the value or from the inner text,  we would need to change Xray. In XML there are limitations for the usage of the `value` attribute, as AFAIK though.

One option here moving forward, in order to make it more agnostic and not have any hardcoding of the annotations that should be either mapped to a value or to inner text, would be to have some sort of explicitly differentiate that whenever using the annotations. However, in playwright we just have the `type` and the `description` fields, correct? 

`
testInfo.annotations.push({ type: 'test_id', description: '1234' });
`

I don't see here a way of signalling it from the outside.. unless we would use a prefix on the `type` or on the  `description` or something like that.
Any ideas?

Comment (dgozman) [reply packages/playwright-test/src/reporters/junit.ts:171]: I see, thank you for clarifying. Let's go with options for now, and provide a nice snippet in the docs:

```js
// JUnit reporter config for XRay.
const xrayOptions = {
  // Whether to add <properties> with all annotations.
  embedAnnotationsAsProperties: true,

  // By default, annotation is reported as <property name='' value=''>.
  // These annotations are reported as <property name=''>value</property>.
  textContentAnnotations: ['test_description'],

  // This will create a "testrun_evidence" property that contains all attachments.
  // Disables [[ATTACHMENT|path]] in the <system-out>.
  embedAttachmentsAsProperty: 'testrun_evidence',

  // Where to put the report.
  outputFile: './xray-report.xml',
};

module.exports = {
  reporter: [['list'],  ['junit', xrayOptions]]
};
```

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [reply packages/playwright-test/src/reporters/junit.ts:179]: I agree this should be disabled by default. I'm inclined for a single global option on the reporter for now, unless someone explicitly asks for more fine-grained control per each test. See my other comment with suggested options.

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [reply packages/playwright-test/src/reporters/junit.ts:185]: Let's go with a single global option for the junit reporter. See my other comment with suggested options.

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [reply packages/playwright-test/src/reporters/junit.ts:296]: Yeah, I am hesitant changing it by default - just in case some of the parsers in the wild breaks and we regress existing users. An option is a safe bet, and we can recommend it in the docs for all new users. A separate PR with this option would be ideal 😄

**Review by dgozman — COMMENTED** [review]
I was on vacations, so didn't get to this before, unfortunately. Sorry!

Let's go with adding some options, and this will be a great change! Note that you'll need to sign CLA to be able to contribute.

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts]: I've made a new PR  for that #12741

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts]: done, updated the PR, including the docs

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts]: done, with a global option for now. updated the PR

**Review by dgozman — COMMENTED** [review]
I think we are on the right track, just a few small details left.

File: docs/src/test-reporters-js.md
Comment (bitcoder) [reply docs/src/test-reporters-js.md:396]: thanks @dgozman  for the feedback, we'll have a look at it

File: tests/playwright-test/reporter-junit.spec.ts
Comment (bitcoder) [reply tests/playwright-test/reporter-junit.spec.ts:309]: right, it shouldn't be there. removed

**Comment by mxschmitt** [comment]
@bitcoder thank you for your contribution! 💯 don't forget to sign the CLA https://github.com/microsoft/playwright/pull/11374#issuecomment-1012083148, otherwise we can't merge PRs.

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:195]: given the previous comment, where I mentioned it would make sense to add the warning to the `<system-err>` element, then I'll keep this approach and the `existsSync` as also exists in the other code block

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [reply packages/playwright-test/src/reporters/junit.ts:197]: Sounds good.

File: packages/playwright-test/src/reporters/junit.ts
Comment (dgozman) [reply packages/playwright-test/src/reporters/junit.ts:195]: Sure, although this would require handling both "does not exist" and "error reading file" situations separately. In the latter case we can even print the error message to `<system-err>`.

File: docs/src/test-reporters-js.md
Comment (bitcoder) [reply docs/src/test-reporters-js.md:449]: sure, makes sense. The branch I have still has a old implementation of testInfo.attach which is not compatible with the latest one. In the doc I'll refer `testInfo.attach`; in the related test I use `testInfo.attachments.push` (and I've left the new testInfo.attach commented..)

**Comment by bitcoder** [comment]
@dgozman  thanks for your patience reviewing this. Hope the latest PR code is closer for merging, but please review it once again.

File: docs/src/test-reporters-js.md
Comment (bitcoder) [reply docs/src/test-reporters-js.md:426]: updated in the latest PR code

File: packages/playwright-test/src/reporters/junit.ts
Comment (bitcoder) [reply packages/playwright-test/src/reporters/junit.ts:196]: I have kept the same login as the existing oneon the initial code, where if file does not exist then it will add it to <system-err> (as it means the user provided a wrong path); if it fails reading the file, it is catch and doesn't do anything. 
This can be changed though, but I would say that then we would also need to change the existing logic on the other code.

**Review by dgozman — APPROVED** [review]
Looks great! I will be able to merge after you sign the CLA - see the comment from microsoft-cla bot.

**Comment by mxschmitt** [comment]
@bitcoder any updates on the CLA? :) (want to reduce risks with merge conflicts)

**Comment by bitcoder** [comment]
hey @mxschmitt , thanks for reaching out. Unfortunately not yet (I have no ETA for it also). I've asked for this to the legal team a few times and it is taking  more than I've anticipated.. :( thanks for your patience in advance

**Comment by dgozman** [comment]
@bitcoder Just a friendly reminder to ping the legal team about this issue 😄 I hope to merge you great PRs!

**Comment by bitcoder** [comment]
> @bitcoder Just a friendly reminder to ping the legal team about this issue 😄 I hope to merge you great PRs!

on it... I also want to move this forward as fast as possible (doing what I can here..)

**Comment by Alje001** [comment]
@bitcoder  @dgozman
Do you have any update on this task? This is a very important feature, which we like to use. 

Thank you for supporting

**Comment by Alje001** [comment]
do you have any update on this task? it is an important feature and we need it for test traceability.

thank you for supporting.

**Review by raghavthind2005 — APPROVED** [review]

**Comment by bitcoder** [comment]
> @bitcoder @dgozman Do you have any update on this task? This is a very important feature, which we like to use.
> 
> Thank you for supporting

not yet, sorry for that. this is taking way longer than I would expect but as Microsoft requires the signed CLA and my company is still going over it, I cannot do anything els right now. Hope for your understanding

**Comment by mxschmitt** [comment]
@bitcoder friendly ping, that you ping your legal team about the CLA, thank you in advance!

**Comment by bitcoder** [comment]
@mxschmitt on it; even talked about that yesterday to upper mgnt

**Comment by Alje001** [comment]
I hope that you @bitcoder will get the Cla soon, because we would like to use this feature

**Comment by bitcoder** [comment]
CLA has been signed; I got the approval from my current company (Xblend).

**Comment by bitcoder** [comment]
@mxschmitt  @dgozman  fyi, this should now be able to be merged

**Pull request merged**
